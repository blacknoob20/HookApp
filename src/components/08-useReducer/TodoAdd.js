import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
    const [{ description }, handleInputChange, reset] = useForm({ description: '' });

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length < 1) return;

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };

        handleAddTodo(newTodo);
        reset();
    };

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form action="">
                <input type="text" name="description" placeholder="Descripcion" autoComplete="off" className="form-control" onChange={handleInputChange} value={description} />
                <button onClick={handleSubmit} className="btn btn-outline-primary btn-lg btn-block mt-2">Agregar</button>
            </form>
        </>
    )
}
