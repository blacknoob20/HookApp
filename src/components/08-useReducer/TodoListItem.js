import React from 'react'

export const TodoListItem = (todo, i, handleToggle, handleDelete) => {
    return (
        <li key={todo.id} className="list-group-item">
            <span onClick={e => handleToggle(todo.id)} className={`${todo.done && 'complete'}`}>{i + 1}. {todo.desc}</span>
            <button onClick={e => handleDelete(todo.id)} className="btn btn-danger mx-3">Borrar</button>
        </li>
    )
}
