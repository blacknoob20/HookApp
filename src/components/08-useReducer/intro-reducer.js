const initState = [{
    id: 1,
    todo: 'Comprar leche',
    done: false,
}];

const todoReducer = (state = initState, action) => {
    if (action?.type === 'agregar') return [...state, action.payload];

    return state;
}

let todos = todoReducer();
const newTodo = {
    id: 2,
    todo: 'Comprar queso',
    done: false,
};

// La accion que se envia al reducer como norma
// debe tener el atributo type
// Opcional
// el atributo payload, que es donde se envia la data
const action = {
    type: 'agregar',
    payload: newTodo

}

todos = todoReducer(todos, action);

console.log(todos);