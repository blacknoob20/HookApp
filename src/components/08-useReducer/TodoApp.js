import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';

import './style.css';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {
    // El dispatch es una funcion que toma una accion y la setea al reducer correspondiente
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleDelete = todoId => {
        console.log(todoId);
        const action = { type: 'del', payload: todoId };

        dispatch(action);
    };

    const handleToggle = todoId => dispatch({ type: 'tog', payload: todoId });

    const handleAddTodo = newTodo => dispatch({ type: 'add', payload: newTodo });

    return (
        <>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="container">
                <div className="row">
                    <div className="col">
                        <TodoList todos={todos} handleToggle={handleToggle} handleDelete={handleDelete} />
                    </div>
                    <div className="col">
                        <TodoAdd handleAddTodo={handleAddTodo} />
                    </div>
                </div>
            </div>
        </>

    )
}
