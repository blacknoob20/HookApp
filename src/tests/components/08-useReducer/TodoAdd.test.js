import { shallow } from 'enzyme';
import { TodoAdd } from '../../../components/08-useReducer/TodoAdd';

describe('Pruebas en el componente <TodoAdd />', () => {
    const handleAddTodo = jest.fn();
    const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

    test('Debe mostrarse correctamente.', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('No debe de llamar al handleAddTodo().', () => {
        const formSubmit = wrapper.find('button').prop('onClick');

        formSubmit({ preventDefault() { } });
        expect(handleAddTodo).toHaveBeenCalledTimes(0);
    });

    test('Debe de llamar la funcio handleAddTodo.', () => {
        const value = 'Aprender Angular';
        wrapper.find('input').simulate('change', { target: { value, name: 'description' } })

        const formSubmit = wrapper.find('button').prop('onClick');

        formSubmit({ preventDefault() { } });
        expect(handleAddTodo).toHaveBeenCalledTimes(1);
        expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object));
        expect(handleAddTodo).toHaveBeenCalledWith({
            id: expect.any(Number),
            desc: value,
            done: false,
        });
        expect(wrapper.find('input').prop('value')).toBe('');
    })


});
