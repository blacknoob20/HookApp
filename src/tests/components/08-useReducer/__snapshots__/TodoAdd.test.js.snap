// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoAdd /> Debe mostrarse correctamente. 1`] = `
<Fragment>
  <h4>
    Agregar TODO
  </h4>
  <hr />
  <form
    action=""
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Descripcion"
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary btn-lg btn-block mt-2"
      onClick={[Function]}
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
